### PR TITLE
Simplify universal impl trait lowering

### LINKED ISSUE #96644: Simplify universal impl-trait collection during lowering
During lowering (`rustc_ast_lowering`), universal impl-trait are desugared to extra generic parameters to the current item, and their bounds added to the where clause.
```rust
fn foo<T>(u: impl Bar) where T: Foo
```
becomes
```rust
fn foo<T, _U>(u: _U) where T: Foo, _U: Bar
```

This is done by carrying two vectors inside `ImplTraitContext::Universal`.
These vectors are only created by `add_implicit_generics` method, and references to them are passed around.
We should prefer having them as fields in `LoweringContext`.

Instructions:
- [ ] Move the two vectors from `ImplTraitContext::Universal` to field in `LoweringContext`;
- [ ] Take and replace the values of these fields in `with_hir_id_owner`, and assert that the value after the call to `f` is empty;
- [ ] Add a `itctx: ImplTraitContext` parameter to `add_implicit_generics` and assert that these fields are empty on return if `itctxt` is not `Universal`;
- [ ] Replace all calls to `lower_generics` by calls to `add_implicit_generics`;
- [ ] Rename `add_implicit_generics` to `lower_generics`;
- [ ] Maybe inline `lower_generics_mut` into `lower_generics` (if this simplifies the logic, not sure).

Please contact me on zulip for any question.

### PR DESCRIPTION
**Pull request by spastorino**
Closes #96644

r? @cjgillot

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -290,11 +293,11 @@ enum ImplTraitPosition {
     ImplReturn,
 }
 
-impl<'a> ImplTraitContext<'_, 'a> {
-    fn reborrow<'this>(&'this mut self) -> ImplTraitContext<'this, 'a> {
+impl ImplTraitContext {
+    fn reborrow<'this>(&'this mut self) -> ImplTraitContext {
```
Comment (fbstj) [inline compiler/rustc_ast_lowering/src/lib.rs]: is `'this` still relevant?

Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Anyway the whole reborrow impl was removed in a later commit.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -290,18 +293,6 @@ enum ImplTraitPosition {
     ImplReturn,
 }
 
-impl<'a> ImplTraitContext<'_, 'a> {
-    fn reborrow<'this>(&'this mut self) -> ImplTraitContext<'this, 'a> {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:294]: :tada:

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 3]
```diff
@@ -479,13 +470,17 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+        let current_impl_trait_defs = std::mem::take(&mut self.impl_trait_defs);
+        let current_impl_trait_bounds = std::mem::take(&mut self.impl_trait_bounds);
 
         // Always allocate the first `HirId` for the owner itself.
         let _old = self.node_id_to_local_id.insert(owner, hir::ItemLocalId::new(0));
         debug_assert_eq!(_old, None);
 
         let item = f(self);
         debug_assert_eq!(def_id, item.def_id());
+        debug_assert!(self.impl_trait_defs.is_empty());
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:482]:
```suggestion
        // `f` should have consumed all the elements in these vectors when constructing `item`.
        debug_assert!(self.impl_trait_defs.is_empty());
```

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 4]
```diff
@@ -1442,21 +1384,17 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
-        mut in_band_ty_params: Option<(
-            NodeId,
-            &mut Vec<hir::GenericParam<'hir>>,
-            &mut Vec<hir::WherePredicate<'hir>>,
-        )>,
+        fn_node_id: Option<NodeId>,
         kind: FnDeclKind,
         make_ret_async: Option<NodeId>,
     ) -> &'hir hir::FnDecl<'hir> {
         debug!(
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1391]: Drive-by: should we use `tracing::instrument` macro here?

File: compiler/rustc_ast_lowering/src/item.rs [hunk 5]
```diff
@@ -1289,11 +1296,23 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
-    ) -> GenericsCtor<'hir> {
+        parent_node_id: NodeId,
+        itctx: ImplTraitContext,
+        f: impl FnOnce(&mut Self) -> T,
+    ) -> (&'hir hir::Generics<'hir>, T) {
+        match itctx {
+            ImplTraitContext::Universal(..) => {}
+            _ => {
+                debug_assert!(self.impl_trait_defs.is_empty());
+                debug_assert!(self.impl_trait_bounds.is_empty());
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/item.rs:1312]: Should these always be empty when entering this function?

File: compiler/rustc_ast_lowering/src/item.rs [hunk 6]
```diff
@@ -1360,22 +1379,34 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+        let extra_lifetimes = self.resolver.take_extra_lifetime_params(parent_node_id);
+        let impl_trait_defs = std::mem::take(&mut self.impl_trait_defs);
+        params.extend(
+            extra_lifetimes
+                .into_iter()
+                .filter_map(|(ident, node_id, res)| {
+                    self.lifetime_res_to_generic_param(ident, node_id, res)
+                })
+                .chain(impl_trait_defs.into_iter()),
+        );
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/item.rs:1397]: Having two `extend` calls may be clearer, those iterators don't give a reliable size estimate anyway.

**Review by cjgillot — COMMENTED** [review]
Thanks @spastorino! I left a few questions.

File: compiler/rustc_ast_lowering/src/item.rs (hunk 5 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/item.rs:1312]: You meant in the not `Universal` case or what? Unsure I got what you meant here.

Comment (cjgillot) [reply compiler/rustc_ast_lowering/src/item.rs:1312]: I mean: should we remove the match and always assert?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 7]
```diff
@@ -243,14 +246,14 @@ pub trait ResolverAstLowering {
...
-enum ImplTraitContext<'b, 'a> {
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ImplTraitContext {
     /// Treat `impl Trait` as shorthand for a new universal generic parameter.
     /// Example: `fn foo(x: impl Debug)`, where `impl Debug` is conceptually
     /// equivalent to a fresh universal parameter like `fn foo<T: Debug>(x: T)`.
     ///
     /// Newly generated parameters should be inserted into the given `Vec`.
-    Universal(&'b mut Vec<hir::GenericParam<'a>>, &'b mut Vec<hir::WherePredicate<'a>>, LocalDefId),
+    Universal(LocalDefId),
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_ast_lowering/src/lib.rs:256]: Ok for follow-up.

[the remark this answers is not in the document]
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:256]: Ok, I've just checked this and yes, it's always equal. We are always creating `Universal` passing `self.current_hir_id_owner` to it.
